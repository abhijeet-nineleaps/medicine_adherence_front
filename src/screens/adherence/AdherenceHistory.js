import React from 'react';
import {
  Text,
  View,
  FlatList,
  PermissionsAndroid,
  Image,
  LogBox,
  Modal,
} from 'react-native';

import Toast from 'react-native-toast-message';
import ProgressCircle from 'react-native-progress-circle';
import {Picker} from '@react-native-picker/picker';
import {Button, Divider} from 'react-native-elements';
import {useFocusEffect} from '@react-navigation/native';
import allreminderdata from '../../components/adherence/allReminderData';
import queryData from '../../repositories/database/queryData';
import * as Progress from 'react-native-progress';
import LottieView from 'lottie-react-native';
import adherence from '../../redux/apis/adherence';
import downloadPdf from '../../components/adherence/downloadPdf';
import MedicinehistoryList from '../../components/organisms/medicineHistoryList';
import globalDb from '../../repositories/database/globalDb';
import Carousel, {Pagination} from 'react-native-snap-carousel';
import styles from './adherenceStyles/AdherenceHistoryStyles';
import Logger from '../../components/logger';

let globalmedId;
LogBox.ignoreLogs(['Require cycle:']);
LogBox.ignoreAllLogs();
let db;
const AdherenceHistory = () => {
  const [pickerValue, setPickerValue] = React.useState();
  const [allreminders, reminders_state] = React.useState([]);
  const [reminder_map_fetched_data, reminder_map_fetched_data_state] =
    React.useState([]);
  const [med_detail, med_detail_state] = React.useState();
  const [sync, syncstate] = React.useState(false);
  const [disableDownload, downloadState] = React.useState(true);
  const [modalVisible, setModalVisible] = React.useState(false);
  const [showDetail, showDetailState] = React.useState(false);
  const [imagearray, setimagearray] = React.useState([]);
  const [index, setindex] = React.useState(0);
  const fetchreminders = async dbs => {
    let reminder_array = [];

    await dbs.transaction(async function (txn) {
      txn.executeSql(
        'CREATE TABLE IF NOT EXISTS User_medicines(user_id INTEGER PRIMARY KEY NOT NULL, medicine_name TEXT, medicine_des TEXT , title TEXT, time TEXT , days TEXT , start_date TEXT , end_date TEXT , status INTEGER , sync INTEGER)',
        [],
      );
      reminder_array = await queryData.getusermeds(txn);
      reminders_state(reminder_array);
    });
  };

  const remindersofparticular_medicine = async med_name => {
    const histoy_obj = await allreminderdata(med_name);
    const output_map = histoy_obj.mapper;
    const {meds_id} = histoy_obj;
    globalmedId = meds_id;
    let f_array = [];
    for (let [key, value] of output_map.entries()) {
      let arr = {date: key, key: {taken: [], not_taken: [], remId: ''}};
      arr.key.taken = value.taken;
      arr.key.not_taken = value.not_taken;
      arr.key.remId = value.remId;
      f_array.push(arr);
    }
    reminder_map_fetched_data_state(f_array);
    let syncData = [];
    f_array.map(mdata => {
      let mobj = {date: '', taken: [], not_taken: [], remId: ''};
      mobj.date = mdata.date;
      mobj.taken = mdata.key.taken;
      mobj.not_taken = mdata.key.not_taken;
      mobj.remId = mdata.key.remId;
      syncData.push(mobj);
    });
    downloadState(true);
    syncstate(true);
    await adherence.syncmedicineHistory({meds_id, syncData});
    syncstate(false);

    downloadState(false);
  };

  const getmed_details = async med_name => {
    await db.transaction(async function (txn) {
      txn.executeSql(
        'SELECT * FROM `User_medicines` WHERE medicine_name = ?',
        [med_name],
        function (_tx, res) {
          med_detail_state(res.rows.item(0));
        },
      );
    });
  };

  useFocusEffect(
    React.useCallback(() => {
      db = globalDb();
      fetchreminders(db);
      return () => {
        /* do nothing */
      };
    }, []),
  );
  const showDetailfun = sDate => {
    if (sDate === null) {
      Toast.show({
        type: 'info',
        text1: 'No Images',
      });
      return;
    }
    setimagearray(sDate);
    showDetailState(true);
    setModalVisible(true);
  };
  async function perm() {
    async () => {
      showDetailState(false);
      await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.WRITE_EXTERNAL_STORAGE,
      );
      await PermissionsAndroid.request(
        PermissionsAndroid.PERMISSIONS.READ_EXTERNAL_STORAGE,
      );
      setModalVisible(true);
      const downloadResp = await downloadPdf(globalmedId);
      setModalVisible(false);
      if (downloadResp !== 'err') {
        Toast.show({
          type: 'success',
          text1: 'Downloaded successfully',
        });
      } else {
        Toast.show({
          type: 'error',
          text1: 'Error while downloading',
        });
      }
    }
  }
  return (
    <View style={styles.container}>
      <Modal
        onRequestClose={() => setModalVisible(false)}
        animationType="fade"
        transparent={true}
        visible={modalVisible}
        style={styles.modal}>
        <View style={styles.carousalView}>
          {showDetail ? (
            <>
              <Carousel
                onSnapToItem={inde => setindex(inde)}
                layout={'stack'}
                data={imagearray}
                renderItem={({item}) => {
                  Logger.loggerInfo('image');
                  return (
                    <View style={styles.carousalImageView}>
                      <Image
                        source={{uri: `${item}`}}
                        resizeMode="contain"
                        style={styles.carousalImage}></Image>
                      <View style={styles.medNameView1}>
                        <Text style={styles.medName}>{pickerValue}</Text>
                      </View>
                    </View>
                  );
                }}
                sliderWidth={660}
                itemWidth={600}></Carousel>
              <Pagination
                dotsLength={imagearray.length}
                activeDotIndex={index}
                containerStyle={styles.paginationContainer}
                dotStyle={styles.paginationDot}
                inactiveDotStyle={styles.paginationInactiveDot}
                inactiveDotOpacity={0.4}
                inactiveDotScale={0.6}
              />
            </>
          ) : (
            <LottieView
              style={{width: 70, height: 70}}
              speed={0.8}
              source={require('../../../assests/animate/generatepdf.json')}
              autoPlay
              loop
            />
          )}
        </View>
      </Modal>
      {sync ? (
        <View style={styles.syncView}>
          <Text style={styles.syncText}>Syncing Data</Text>
          <Progress.CircleSnail
            spinDuration={400}
            size={30}
            color={['white']}
          />
        </View>
      ) : (
        <></>
      )}
      <View style={styles.conatiner1}>
        <View style={styles.medNameView}>
          <Picker
            mode="dropdown"
            style={styles.medNamePicker}
            selectedValue={pickerValue}
            onValueChange={itemValue => {
              setPickerValue(itemValue);
              remindersofparticular_medicine(itemValue);
              getmed_details(itemValue);
            }}>
            {allreminders
              .filter(it => it.status === 1)
              .map(it => {
                return (
                  <Picker.Item
                    key={it.medicine_name}
                    label={it.medicine_name}
                    value={it.medicine_name}
                  />
                );
              })}
          </Picker>
        </View>
      </View>
      <View style={styles.container2}>
        <View style={styles.container2Left}>
          <Text style={styles.container2LeftText}>Overall Performance </Text>
        </View>
        <View style={styles.container2Right}>
          <ProgressCircle
            percent={
              med_detail &&
              Math.round(
                (med_detail.current_count / med_detail.total_med_reminders) *
                  100,
              )
            }
            radius={35}
            borderWidth={3}
            color="#4dd0e1"
            bgColor="#fff">
            <Text style={styles.container2RightText}>
              {med_detail &&
                Math.round(
                  (med_detail.current_count / med_detail.total_med_reminders) *
                    100,
                ) + '%'}
            </Text>
          </ProgressCircle>
        </View>
      </View>
      <Divider />
      <View style={styles.conatiner3}>
        <Text style={styles.conatiner3Text}> Detailed Report</Text>
      </View>
      {
        <FlatList
          data={reminder_map_fetched_data}
          renderItem={({item}) => {
            return (
              <MedicinehistoryList
                item={item}
                medName={pickerValue}
                showimgfun={showDetailfun}></MedicinehistoryList>
            );
          }}></FlatList>
      }
      <Button
        disabled={disableDownload}
        title="Download PDF"
        buttonStyle={styles.button}
        id='perm'
        onPress={perm}></Button>
    </View>
  );
};

export default AdherenceHistory;
