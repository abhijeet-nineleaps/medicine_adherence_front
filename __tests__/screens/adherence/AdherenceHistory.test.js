import React from 'react';
import {create} from 'react-test-renderer';
import AdherenceHistory from '../../../src/screens/adherence/AdherenceHistory';
import Enzyme from 'enzyme';
import {shallow} from 'enzyme';
import Adapter from '@wojtekmaj/enzyme-adapter-react-17';
Enzyme.configure({adapter: new Adapter()});
jest.mock('@react-navigation/native', () => ({
  ...jest.requireActual('@react-navigation/native'),
  useFocusEffect: jest.fn(),
}));
jest.mock("rn-fetch-blob", () => ({
  default: jest.fn(),
}));
describe('Settins Screen', () => {
  it('renders correctly', () => {
    const tree = create(<AdherenceHistory />);
    expect(tree).toMatchSnapshot();
  });
  it('test open save button', () => {
    const mockFn = jest.fn();
    const wrapper = shallow(<AdherenceHistory perm={mockFn} />);
    wrapper.find('#perm').simulate('press');
  });
});
