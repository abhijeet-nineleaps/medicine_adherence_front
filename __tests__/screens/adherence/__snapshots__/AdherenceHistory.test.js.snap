// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Settins Screen renders correctly 1`] = `
<View
  style={
    Object {
      "backgroundColor": "white",
      "height": "100%",
    }
  }
>
  <Modal
    animationType="fade"
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    style={
      Object {
        "backgroundColor": "white",
      }
    }
    transparent={true}
    visible={false}
  />
  <View
    style={
      Object {
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "borderColor": "lightgrey",
          "borderEndWidth": 1,
          "width": "100%",
        }
      }
    >
      <View
        style={
          Object {
            "backgroundColor": "white",
            "borderColor": "lightgrey",
            "borderWidth": 3,
            "color": "black",
            "elevation": 3,
          }
        }
      >
        <RNCPicker
          items={Array []}
          numberOfLines={1}
          onChange={[Function]}
          selectedIndex={0}
          style={
            Array [
              Object {
                "height": 216,
              },
              undefined,
            ]
          }
        />
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "padding": 7,
      }
    }
  >
    <View
      style={
        Object {
          "left": 10,
        }
      }
    >
      <Text
        style={
          Object {
            "color": "black",
            "fontSize": 18,
            "marginTop": 30,
          }
        }
      >
        Overall Performance 
      </Text>
    </View>
    <View
      style={
        Object {
          "alignItems": "center",
          "margin": 10,
          "paddingRight": 20,
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "alignItems": "center",
              "justifyContent": "center",
            },
            Object {
              "backgroundColor": "#999",
              "borderRadius": 35,
              "height": 70,
              "width": 70,
            },
          ]
        }
      >
        <View
          style={
            Array [
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              },
              Object {
                "height": 70,
                "width": 35,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomRightRadius": 0,
                  "borderTopRightRadius": 0,
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                },
                Object {
                  "backgroundColor": "#4dd0e1",
                  "borderRadius": 35,
                  "height": 70,
                  "overflow": "hidden",
                  "transform": Array [
                    Object {
                      "translateX": 17.5,
                    },
                    Object {
                      "rotate": "NaNdeg",
                    },
                    Object {
                      "translateX": -17.5,
                    },
                  ],
                  "width": 35,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "left": 0,
                "position": "absolute",
                "top": 0,
              },
              Object {
                "height": 70,
                "width": 35,
              },
            ]
          }
        >
          <View
            style={
              Array [
                Object {
                  "borderBottomRightRadius": 0,
                  "borderTopRightRadius": 0,
                  "left": 0,
                  "position": "absolute",
                  "top": 0,
                },
                Object {
                  "backgroundColor": "#999",
                  "borderRadius": 35,
                  "height": 70,
                  "overflow": "hidden",
                  "transform": Array [
                    Object {
                      "translateX": 17.5,
                    },
                    Object {
                      "rotate": "0deg",
                    },
                    Object {
                      "translateX": -17.5,
                    },
                  ],
                  "width": 35,
                },
              ]
            }
          />
        </View>
        <View
          style={
            Array [
              Object {
                "alignItems": "center",
                "justifyContent": "center",
                "overflow": "hidden",
              },
              Object {
                "backgroundColor": "#fff",
                "borderRadius": 32,
                "height": 64,
                "width": 64,
              },
            ]
          }
        >
          <Text
            style={
              Object {
                "color": "#4dd0e1",
                "fontSize": 18,
              }
            }
          />
        </View>
      </View>
    </View>
  </View>
  <View
    style={
      Object {
        "borderBottomColor": "#bcbbc1",
        "borderBottomWidth": 0.5,
      }
    }
  />
  <View
    style={
      Object {
        "backgroundColor": "lightgrey",
        "marginBottom": 5,
        "padding": 15,
      }
    }
  >
    <Text
      style={
        Object {
          "fontWeight": "600",
        }
      }
    >
       Detailed Report
    </Text>
  </View>
  <RCTScrollView
    data={Array []}
    getItem={[Function]}
    getItemCount={[Function]}
    keyExtractor={[Function]}
    onContentSizeChange={[Function]}
    onLayout={[Function]}
    onMomentumScrollBegin={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    removeClippedSubviews={false}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    viewabilityConfigCallbackPairs={Array []}
  >
    <View />
  </RCTScrollView>
  <View
    style={
      Array [
        Object {
          "borderRadius": 3,
          "overflow": "hidden",
        },
        Object {
          "borderRadius": 3,
        },
        undefined,
        false,
      ]
    }
  >
    <View
      accessibilityRole="button"
      accessibilityState={
        Object {
          "busy": false,
          "disabled": true,
        }
      }
      accessible={true}
      collapsable={false}
      focusable={true}
      onClick={[Function]}
      onResponderGrant={[Function]}
      onResponderMove={[Function]}
      onResponderRelease={[Function]}
      onResponderTerminate={[Function]}
      onResponderTerminationRequest={[Function]}
      onStartShouldSetResponder={[Function]}
      style={
        Object {
          "opacity": 1,
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "hsl(208, 8%, 90%)",
            "borderColor": "#2089dc",
            "borderRadius": 3,
            "borderWidth": 0,
            "flexDirection": "row",
            "justifyContent": "center",
            "padding": 8,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "hsl(208, 8%, 63%)",
              "fontSize": 18,
              "paddingVertical": 1,
              "textAlign": "center",
            }
          }
        >
          Download PDF
        </Text>
      </View>
    </View>
  </View>
</View>
`;
